### PR TITLE
Use configuration from ProviderSettings in OAuth2AuthorizationServerConfigurer

### LINKED ISSUE #182: Use ProviderSettings in OAuth2AuthorizationServerConfigurer
Hello,

Happy new year !

I probably found a bug in OAuth2AuthorizationServerConfigurer class. 
The tokenEndpoint and tokenRevocationEndpoint properties from ProviderSettings are not used in OAuth2ClientAuthenticationFilter for :
- creation of OAuth2ClientAuthenticationFilter
- creation of DelegatingAuthenticationEntryPoint

As a result, using custom urls for both endpoints doesn't work at all.

Thank you for your support.

Flo

### PR DESCRIPTION
**Pull request by florianberthe**
Fix #182 issue

### CODE REVIEW

**Comment by pivotal-issuemaster** [comment]
@florianberthe Please sign the [Contributor License Agreement](https://cla.pivotal.io/sign/spring?repositoryId=spring-projects-experimental/spring-authorization-server&pullRequestId=184)!

[Click here](https://cla.pivotal.io/sync/spring?repositoryId=spring-projects-experimental/spring-authorization-server&pullRequestId=184) to manually synchronize the status of this Pull Request.

See the [FAQ](https://cla.pivotal.io/about) for frequently asked questions.

**Comment by pivotal-issuemaster** [comment]
@florianberthe Thank you for signing the [Contributor License Agreement](https://cla.pivotal.io/sign/spring?repositoryId=spring-projects-experimental/spring-authorization-server&pullRequestId=184)!

File: oauth2-authorization-server/src/main/java/org/springframework/security/config/annotation/web/configurers/oauth2/server/authorization/OAuth2AuthorizationServerConfigurer.java [hunk 1]
```diff
@@ -196,7 +196,9 @@ public void init(B builder) {
 		if (exceptionHandling != null) {
 			LinkedHashMap<RequestMatcher, AuthenticationEntryPoint> entryPoints = new LinkedHashMap<>();
 			entryPoints.put(
-					new OrRequestMatcher(this.tokenEndpointMatcher, this.tokenRevocationEndpointMatcher),
+					new OrRequestMatcher(
```
Comment (jgrandja) [inline oauth2-authorization-server/src/main/java/org/springframework/security/config/annotation/web/configurers/oauth2/server/authorization/OAuth2AuthorizationServerConfigurer.java:199]: This change leaves `this.tokenEndpointMatcher` and `this.tokenRevocationEndpointMatcher` in an inconsistent state. The members should be updated.

**Review by jgrandja — CHANGES_REQUESTED** [review]
Thanks for the PR @florianberthe. Please see review comment.

Also, please add test(s) as this is required for any change that is submitted.

File: oauth2-authorization-server/src/main/java/org/springframework/security/config/annotation/web/configurers/oauth2/server/authorization/OAuth2AuthorizationServerConfigurer.java (hunk 1 above)
Comment (florianberthe) [reply oauth2-authorization-server/src/main/java/org/springframework/security/config/annotation/web/configurers/oauth2/server/authorization/OAuth2AuthorizationServerConfigurer.java:199]: @jgrandja I know but all other matchers are also in an inconsistent state. I can change all matchers if you agree.

Comment (jgrandja) [reply oauth2-authorization-server/src/main/java/org/springframework/security/config/annotation/web/configurers/oauth2/server/authorization/OAuth2AuthorizationServerConfigurer.java:199]: Yes, please change all matchers and anything else that applies within `OAuth2AuthorizationServerConfigurer`.

For the tests, take a look at the existing integration tests in package `org.springframework.security.config.annotation.web.configurers.oauth2.server.authorization`

Comment (florianberthe) [reply oauth2-authorization-server/src/main/java/org/springframework/security/config/annotation/web/configurers/oauth2/server/authorization/OAuth2AuthorizationServerConfigurer.java:199]: I will try to change it before the end of the week.

**Comment by florianberthe** [comment]
@jgrandja I pushed changes for all matchers. Is that what you thought about ?

File: oauth2-authorization-server/src/main/java/org/springframework/security/oauth2/server/authorization/config/ProviderSettings.java [hunk 2]
```diff
@@ -146,10 +151,10 @@ public ProviderSettings tokenRevocationEndpoint(String tokenRevocationEndpoint)
 
 	protected static Map<String, Object> defaultSettings() {
 		Map<String, Object> settings = new HashMap<>();
-		settings.put(AUTHORIZATION_ENDPOINT, "/oauth2/authorize");
-		settings.put(TOKEN_ENDPOINT, "/oauth2/token");
-		settings.put(JWK_SET_ENDPOINT, "/oauth2/jwks");
-		settings.put(TOKEN_REVOCATION_ENDPOINT, "/oauth2/revoke");
+		settings.put(AUTHORIZATION_ENDPOINT, OAuth2AuthorizationEndpointFilter.DEFAULT_AUTHORIZATION_ENDPOINT_URI);
```
Comment (jgrandja) [inline oauth2-authorization-server/src/main/java/org/springframework/security/oauth2/server/authorization/config/ProviderSettings.java:154]: Please revert the changes in `ProviderSettings` as it should not depend on `OAuth2AuthorizationEndpointFilter` or the other `Filter`'s

**Review by jgrandja — CHANGES_REQUESTED** [review]
`initEndpointMatchers()` is along the right path, however, this change is causing multiple failing tests. Likely because a NPE is happening when `authorizationServerConfigurer			.getEndpointMatchers()` is called [here](https://github.com/spring-projects-experimental/spring-authorization-server/blob/d97235d0bbf660f81d5399a2a1c52b92f782ace3/oauth2-authorization-server/src/main/java/org/springframework/security/config/annotation/web/configuration/OAuth2AuthorizationServerConfiguration.java#L49).

FYI, I was planning on updating this bit of code closer to the `0.1.0` release as I knew it would be tricky. There is a timing issue happening here on when the matchers are initialized (from `ApplicationContext` start) vs. when `authorizationServerConfigurer			.getEndpointMatchers()` is called.

My initial thought on a possible solution is that `authorizationServerConfigurer			.getEndpointMatchers()` would need to be changed to return a singular `RequestMatcher` that is initialized when it's first called in order for the `initEndpointMatchers()` to be called first. I think this might work.

File: oauth2-authorization-server/src/main/java/org/springframework/security/oauth2/server/authorization/config/ProviderSettings.java (hunk 2 above)
Comment (florianberthe) [reply oauth2-authorization-server/src/main/java/org/springframework/security/oauth2/server/authorization/config/ProviderSettings.java:154]: Regarding this change, default provider settings must be initialized with filter's default value. Don't you think ? Maybe default endpoint URI in filters become useless.

**Comment by florianberthe** [comment]
Good news! With latest change, all tests passed now.

File: oauth2-authorization-server/src/main/java/org/springframework/security/oauth2/server/authorization/config/ProviderSettings.java (hunk 2 above)
Comment (florianberthe) [reply oauth2-authorization-server/src/main/java/org/springframework/security/oauth2/server/authorization/config/ProviderSettings.java]: I rollbacked this change

File: oauth2-authorization-server/src/main/java/org/springframework/security/config/annotation/web/configurers/oauth2/server/authorization/OAuth2AuthorizationServerConfigurer.java [hunk 3]
```diff
@@ -80,21 +78,20 @@
...
-			JwkSetEndpointFilter.DEFAULT_JWK_SET_ENDPOINT_URI, HttpMethod.GET.name());
-	private final RequestMatcher oidcProviderConfigurationEndpointMatcher = new AntPathRequestMatcher(
-			OidcProviderConfigurationEndpointFilter.DEFAULT_OIDC_PROVIDER_CONFIGURATION_ENDPOINT_URI, HttpMethod.GET.name());
+	private RequestMatcher authorizationEndpointMatcher;
+	private RequestMatcher tokenEndpointMatcher;
+	private RequestMatcher tokenRevocationEndpointMatcher;
+	private RequestMatcher jwkSetEndpointMatcher;
+	private RequestMatcher oidcProviderConfigurationEndpointMatcher;
+	private final RequestMatcher endpointMatchers = request -> {
+		RequestMatcher matcher = new OrRequestMatcher(
```
Comment (jgrandja) [inline oauth2-authorization-server/src/main/java/org/springframework/security/config/annotation/web/configurers/oauth2/server/authorization/OAuth2AuthorizationServerConfigurer.java:87]: An `OrRequestMatcher` will be created for each `request`. To avoid this, please change the lambda as follows:

`this.authorizationEndpointMatcher.matches(request) || this.tokenEndpointMatcher.matches(request) || ...`

**Review by jgrandja — CHANGES_REQUESTED** [review]
Looks good @florianberthe !

Please see the additional review comments. Also, please add tests that demonstrate that the configuration from the `ProviderSettings` is actually used. See [comment](https://github.com/spring-projects-experimental/spring-authorization-server/pull/184#discussion_r553502919) for examples.

**Comment by florianberthe** [comment]
@jgrandja please review latest changes

**Comment by florianberthe** [comment]
@jgrandja I updated my code further to you latest changes (Removal of CryptoSource). Please note that `requestWhenJwkSetValidThenReturnKeys` test is failing now. 
I think there is an issue with `this.jwkSelector = new JWKSelector(new JWKMatcher.Builder().publicOnly(true).build());` in `NimbusJwkSetEndpointFilter`

File: oauth2-authorization-server/src/test/java/org/springframework/security/config/annotation/web/configurers/oauth2/server/authorization/NimbusJwkSetRetrievalTests.java [hunk 4]
```diff
@@ -0,0 +1,139 @@
...
+
+		this.mvc.perform(get(NimbusJwkSetEndpointFilter.DEFAULT_JWK_SET_ENDPOINT_URI))
+				.andExpect(status().isOk())
+				.andExpect(header().string(HttpHeaders.CACHE_CONTROL, containsString("no-store")))
+				.andExpect(header().string(HttpHeaders.PRAGMA, containsString("no-cache")))
+				.andExpect(jsonPath("$.keys").isNotEmpty())
+				.andExpect(jsonPath("$.keys").isArray());
+
+	}
+
```
Comment (florianberthe) [inline oauth2-authorization-server/src/test/java/org/springframework/security/config/annotation/web/configurers/oauth2/server/authorization/NimbusJwkSetRetrievalTests.java:92]: `requestWhenJwkSetValidThenReturnKeys` test is failing now. 
I think there is an issue with `this.jwkSelector = new JWKSelector(new JWKMatcher.Builder().publicOnly(true).build());` in `NimbusJwkSetEndpointFilter`

Comment (florianberthe) [reply oauth2-authorization-server/src/test/java/org/springframework/security/config/annotation/web/configurers/oauth2/server/authorization/NimbusJwkSetRetrievalTests.java:92]: See #198 198

File: oauth2-authorization-server/src/test/java/org/springframework/security/config/annotation/web/configurers/oauth2/server/authorization/NimbusJwkSetRetrievalTests.java [hunk 5]
```diff
@@ -0,0 +1,139 @@
...
+import com.nimbusds.jose.jwk.JWKSet;
+import com.nimbusds.jose.jwk.source.JWKSource;
+import com.nimbusds.jose.proc.SecurityContext;
+
+/**
+ * Integration tests for the JWK Set requests.
+ *
+ * @author Florian Berthe
+ */
+public class NimbusJwkSetRetrievalTests {
```
Comment (jgrandja) [inline oauth2-authorization-server/src/test/java/org/springframework/security/config/annotation/web/configurers/oauth2/server/authorization/NimbusJwkSetRetrievalTests.java:53]: Let's rename to `JwkSetTests`

File: oauth2-authorization-server/src/main/java/org/springframework/security/config/annotation/web/configurers/oauth2/server/authorization/OAuth2AuthorizationServerConfigurer.java [hunk 6]
```diff
@@ -137,21 +132,19 @@
...
-	 * @return a {@code List} of {@link RequestMatcher}'s for the authorization server endpoints
+	 * @return a {@link RequestMatcher} for the authorization server endpoints
 	 */
-	public List<RequestMatcher> getEndpointMatchers() {
-		// TODO Initialize matchers using URI's from ProviderSettings
-		return Arrays.asList(this.authorizationEndpointMatcher, this.tokenEndpointMatcher,
-				this.tokenRevocationEndpointMatcher, this.jwkSetEndpointMatcher,
-				this.oidcProviderConfigurationEndpointMatcher);
+	public RequestMatcher getEndpointsMatcher() {
+		return endpointsMatcher;
```
Comment (jgrandja) [inline oauth2-authorization-server/src/main/java/org/springframework/security/config/annotation/web/configurers/oauth2/server/authorization/OAuth2AuthorizationServerConfigurer.java:140]: Prefix `this`

File: oauth2-authorization-server/src/test/java/org/springframework/security/config/annotation/web/configurers/oauth2/server/authorization/NimbusJwkSetRetrievalTests.java [hunk 7]
```diff
@@ -0,0 +1,139 @@
+/*
+ * Copyright 2020 the original author or authors.
```
Comment (jgrandja) [inline oauth2-authorization-server/src/test/java/org/springframework/security/config/annotation/web/configurers/oauth2/server/authorization/NimbusJwkSetRetrievalTests.java:2]: Update to 2020-2021

File: oauth2-authorization-server/src/test/java/org/springframework/security/config/annotation/web/configurers/oauth2/server/authorization/OAuth2AuthorizationCodeGrantTests.java [hunk 8]
```diff
@@ -112,6 +114,7 @@ public static void init() {
 		jwtEncoder = new NimbusJwsEncoder(jwkSource);
 		jwtCustomizer = mock(BiConsumer.class);
 		jwtEncoder.setJwtCustomizer(jwtCustomizer);
+		providerSettings = new ProviderSettings().authorizationEndpoint("/test/authorize");
```
Comment (jgrandja) [inline oauth2-authorization-server/src/test/java/org/springframework/security/config/annotation/web/configurers/oauth2/server/authorization/OAuth2AuthorizationCodeGrantTests.java:117]: Please add tests for token endpoint by setting `.tokenEndpoint("/test/token");`

File: oauth2-authorization-server/src/test/java/org/springframework/security/config/annotation/web/configurers/oauth2/server/authorization/OAuth2ClientCredentialsGrantTests.java [hunk 9]
```diff
@@ -140,6 +143,38 @@ public void requestWhenTokenRequestPostsClientCredentialsThenTokenResponse() thr
 		verify(authorizationService).save(any());
 	}
 
+	@Test
+	public void requestWhenCustomProviderSettingsThenOk() throws Exception {
```
Comment (jgrandja) [inline oauth2-authorization-server/src/test/java/org/springframework/security/config/annotation/web/configurers/oauth2/server/authorization/OAuth2ClientCredentialsGrantTests.java:147]: Let's remove these tests since it will be covered in `OAuth2AuthorizationCodeGrantTests` (see previous comment)

File: oauth2-authorization-server/src/test/java/org/springframework/security/config/annotation/web/configurers/oauth2/server/authorization/OAuth2RefreshTokenGrantTests.java [hunk 10]
```diff
@@ -127,6 +131,48 @@ public void requestWhenRefreshTokenRequestValidThenReturnAccessTokenResponse() t
 
 	}
 
+	@Test
+	public void requestWhenCustomProviderSettingsThenOk() throws Exception {
```
Comment (jgrandja) [inline oauth2-authorization-server/src/test/java/org/springframework/security/config/annotation/web/configurers/oauth2/server/authorization/OAuth2RefreshTokenGrantTests.java:135]: Let's remove these tests since it will be covered in `OAuth2AuthorizationCodeGrantTests` (see previous comment)

**Review by jgrandja — CHANGES_REQUESTED** [review]
Thanks for the updates @florianberthe !

There are a couple more minor changes required and then we should be able to merge.

On next update, please squash to a single commit based on master. Thanks!

**Comment by florianberthe** [comment]
@jgrandja all done 👍

**Comment by jgrandja** [comment]
@florianberthe There are 18 commits in this PR - looks like you did a merge. Can you please [rebase on top of master](https://www.atlassian.com/git/tutorials/merging-vs-rebasing). This should result in 1 commit on top of latest master.
